[Woody] 미션 2: 체스판 생성

jihye-woo opened the pull request:
안녕하세요 Woody 입니다!  미션 2 PR 드립니다!

### ☝ 리뷰 전 말씀드리고 싶은 부분

- **Color 변수 타입**
  - enum class → static final ( Pawn 클래스 변수 ) 
  - 원래 Pawn의 color 정보를 담기 위해 Color enum을 구현해주었지만, 이전 코드리뷰 #29 를 반영하고 미션2의 요구사항에 최대한 맞추기 위해서 Pawn의 일반 클래스 상수로 변경하였습니다.

- **컴파일 에러 테스트 케이스**
  - 미션 요구사항에 맞춰 컴파일 에러가 발생하는 테스트 메소드를 작성하였습니다. 미션을 수행했다는 의미에서 컴파일 에러가 발생하는 코드임에도 불구하고 그대로 커밋하였는데 감안해서 리뷰해주시면 감사드리겠습니다!

### 📃 구현 과정 - 코드를 이해하시는데 도움이 될까 싶어서 올립니다 (안보시고 넘어가주셔도 괜찮습니다!)
- **Pawn 클래스**
   - Pawn 기본 생성자 구현
   
- **Color 상수화**
   - 처음에는 enum class로 구현했다가 이후 Pawn의 클래스 상수로 변경
   
- **BoardTest 클래스 구현**
   - Board에 Pawn이 정상적으로 추가되는지 테스트
   
- **Board 클래스 구현**
    - add() - 체스말 추가
    - size() - 체스말 갯수
    - findPawn() - 인덱스로 체스말 가져오기
    - ✔ 현재는 모든 메소드가 체스말이 아닌 Pawn에 의존적인 상태입니다.

 - **pieces 패키지 추가 후 분리**
   - 체스 말에 해당하는 Pawn과 테스트 클래스 PawnTest은 pieces 패키지로 이동
   

새해 복 많이 받으세요🎊 감사합니다🙇‍♀️

jihye-woo committed:
docs : 미션2 체크리스트 추가

  > 미션2.md 추가

jihye-woo committed:
feat : Pawn 기본 생성자 구현

  > Pawn 기본 생성자
    - "white" 값을 가진 Pawn 인스턴스 생성

  > PawnTest 테스트 메소드 추가
    - Pawn 기본 생성자 관련 테스트 메소드 추가
    - create_기본생성자()
    - 테스트 통과

jihye-woo committed:
feat : Color enum 구현

  > Color enum
    - white, black 값

jihye-woo committed:
edit : Color enum 적용

  > PawnTest
     - color 관련 String 변수 Color 타입으로 변경
     - 테스트 통과

  > Pawn
     - color 필드 값 타입 변경
     - String → Color enum

jihye-woo committed:
refactor : 코드리뷰 반영 (minor issues)

  > enum 상수 대문자 변경
    - "white" "black" → "WHITE" "BLACK"

  > encoding 설정
    - encodings.xml 파일 추가
    - build.gradle 인코딩 설정 코드 삭제

  > PawnTest 메소드 이름
    - 좀 더 구체적으로 메소드 이름 변경

jihye-woo committed:
refactor : 코드리뷰 반영 (Pawn color 타입 변경)

  > Pawn Color 타입 : Color enum → 문자열 상수
    - Color enum 클래스 삭제
    - Pawn 클래스 상수로 WHITE_COLOR("white")와 BLACK_COLOR("black") 선언

  > 테스트 메소드 public 접근지시자 제거
    - JUnit5부터 @Test, 접근지시자 유무와 상관없이 실행가능

jihye-woo committed:
feat: BoardTest 클래스 생성

> 빈 BoardTest 클래스 생성
  - 빈 create() 테스트 메소드 생성
  - 테스트 성공

jihye-woo committed:
feat: create() 테스트 메소드 생성

> 빈 Board 클래스 생성
  - Board 클래스에 대한 테스트 메소드 구현을 위해
  - 빈 add(), size(), findPawn() 생성

> 테스트 실행 후 test failed 확인

jihye-woo committed:
edit : Board 클래스 수정

> Board 클래스 수정
  - field : Pawn을 담는 ArrayList 추가
  - constructor : ArrayList<Pawn> 초기화
  - method : add(), size(), findPawn() 수정

> 테스트 실행 후 test passed 확인

> 테스트 메소드 public 접근지시자 제거

jihye-woo committed:
edit : 패키지 분리

> pieces 패키지 생성
  - Pawn과 PawnTest를 pieces 아래에 위치

> 테스트 실행 후 test passed 확인

jihye-woo committed:
feat : Pawn이 아닌 객체 Board에 추가 - 컴파일 에러

> 테스트 메소드 생성
  - 폰이 아닌 객체 Board에 추가하는 테스트 메소드 작성
  - 컴파일 에러 확인

> minor fixes
  - 테스트 메소드 이름 구체적으로 변경
  - Pawntest에서 assertThrows()에 영향을 주지 않는 코드는 assertThrows() 밖으로 위치

jihye-woo committed:
refactor : BoardTest 중복 제거

> BoardTest 중복 제거
 - addPawnAndCheckTheBoard()
   - 폰을 추가하고 업데이트 하는 중복 로직 분리
 - setup() @BeforeEach
   - 가장 먼저 보드를 생성하는 공통 부분을 BeforeEach를 사용해서 분리

jihye-woo committed:
fix : BoardTest 테스트 제거

> 컴파일 에러 발생하는 테스트 메소드 제거
 - add_보드위에_폰이_아닌_객체추가() 삭제

jihye-woo commented on .idea/encodings.xml:
```diff
@@ -0,0 +1,7 @@
+    <file url="file://$PROJECT_DIR$/src/test/java/net/honux/pieces/MainTest.java" charset="UTF-8" />
+    <file url="PROJECT" charset="UTF-8" />
+  </component>
+</project>
```
엣 분명 자동 reformat 기능을 추가했는데도 아직 통행금지 표시가.. 떴네요ㅠㅠ 따로 엔터 눌러줘야겠습니다😂😂

ksundong replied:
@jihye-woo 
요 파일은 굳이 커밋되지 않아도 되는 파일입니다. `gitignore`에 대해서 학습해보세요!

polynomeer commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,32 @@
+    @DisplayName("보드 위에 폰 추가하고 추가된 폰이 보드에 정상적으로 반영되어야 한다.")
+    void add_폰이_보드에_정상적으로_추가되는지_확인() throws Exception {
+        Pawn white = new Pawn(Pawn.WHITE_COLOR);
+        addPawnAndCheckTheBoard(white);
+
+        Pawn black = new Pawn(Pawn.BLACK_COLOR);
+        addPawnAndCheckTheBoard(black);
+    }
+
+    private void addPawnAndCheckTheBoard(Pawn pawn) {
+        int sizeBeforeAddThePawn = board.size();
+        board.add(pawn);
+        Assertions.assertEquals(sizeBeforeAddThePawn + 1, board.size());
+        Assertions.assertEquals(pawn, board.findPawn(sizeBeforeAddThePawn));
```
안녕하세요, 혹시 `@Test`의 메서드를 제외하고 변수명과 메서드명을 길게 적는 이유가 있나요? 테스트 코드에서는 모든 메서드명을 길게 적는게 관습인지는 잘 모르겠지만, 혹시 그런 이유가 있다면 알려주시면 감사하겠습니다.

그래서 저는 그냥 개인적인 의견 입니다만, 변수명과 메서드명이 너무 긴것이 아닌가 생각이 들어서요! 설연휴 잘 보내세요~ :)

kyupid replied:
@Polynomeer

제가 [디온으로부터 미션 1에서 받은 리뷰](https://github.com/codesquad-members-2021/java-chess/pull/37#discussion_r574397748)에서 참고한 제 개인적인 의견은,

> 메서드 명도 보다 자세했음 좋겠습니다!

라고 언급하신 것을 보아, 테스트 메서드의 `@DisplayName` 이나 메서드 명은 중복되지 않는 선에서 최대한 자세하게 적어주는 게 좋다고 생각했습니다.

kyupid replied:
앗 죄송합니다.

질문을 잘못읽었습니다. 테스트 메서드를 **제외한**  변수명과 메서드명이었군요. 😥

jihye-woo replied:
@Polynomeer 안녕하세요! 
제가 네이밍을 할 때 '나중에 내가 이 코드를 봤을 때 바로 이해할 수 있을까'라는 점을 먼저 생각하려는 편인데요! 의도를 분명하게 밝히는 것을 우선시 하다보니 가독성을 신경쓰지 못했습니당.. 
### 변수명
`Assertions.assertEquals(size + 1, board.size());` 이런 식으로 작성하면 코드를 바로 이해하기는 힘들 것 같아서 변수에다가 구체적인 정보를 넣다보니 저렇게 길어져버렸네요😂
### 함수명
클린코드 제 3장 함수편을 읽다가 
> '길고 서술적인 이름이 짧고 어려운 이름보다 좋다. 길고 서술적인 이름이 길고 서술적인 주석보다 좋다'

라는 말을 보고서, 네이밍을 할 때 최대한 함수가 하는 일이 드러나도록 작성하는게 우선순위가 되었네요...! 하지만 마찬가지로 너무 길거나 구체적이라는 느낌이 들면 최대한 알아보기 쉽게 더 좋은 이름을 찾아봐야겠다는 생각이 듭니다 좋은 의견 주셔서 감사합니다!!😊😊

jihye-woo replied:
@kyu-kim-kr도 안녕하세요!ㅎㅎ 항상 코멘트 달아주셔서 감사드려요! 큐 덕분에 저도 좀 더 열심히 코드리뷰에 참여해야겠다는 마음이 생기네요! 감사합니다😊😊

@kyu-kim-kr  @Polynomeer 모두들 설 연휴 잘 보내세요! 감사합니다🙇‍♀️

ksundong replied:
좋은 의견이네요. 저는 메서드 이름의 길이보다는 얼마나 이 메서드가 하는 일을 잘 설명하는가가 더 중요하다고 생각합니다.
그리고, 메서드 이름이 길어진다면 혹시 메서드가 너무 많은 일을 하는 것은 아닌지 의심해 볼 수도 있다고 생각해요.
이건 사람마다 주관적인 부분이 있는거라서, 나중에 협업 할 때는 다른 팀원과 얘기를 나눠봐도 좋을 것 같아요!

polynomeer replied:
@kyu-kim-kr @jihye-woo @ksundong 친절한 답변 감사합니다! 사실 저도 다소 길더라도 명확한 메서드명에 낫다고 생각하는 쪽입니다. 하지만 Dion이 말한대로 동사가 두 개나 들어가는 부분이 있어서, 혹시 두 가지 이상의 일을 하는게 아닌가 생각했습니다. 이렇게 토론해보니 얻는 것이 많네요!

honux77 replied:
ㅎㅎㅎ 잘 읽었습니다. 변수명은 읽기 좋게 짓는 것이 좋습니다. 근데 그 부분은 사람마다 팀마다 다를 수 있겠죠. 
저는 for 문에서는 i를 선호하는데 회사나 팀에 따라 이 경우 reject되는 경우도 있습니다. 

경험이 쌓이면 구체적인 의미 전달과 지나치게 길어서 가독성을 해치는 경우 사이의 밸런스를 잘 맞추게 될 것 같아요.

jihye-woo replied:
코멘트 감사드려요! 두 가지 일을 하는 코드를 최대한 동사 단위로 분할해보려 했는데 두 assertion 메소드들이 공유하고 있는 상태가 많아서, 오히려 전체 코드 길이가 길어지더라구요..!

그래서 #44 에서 August의 코드를 참고해 보았는데, 예상되는 index를 직접적으로 넣어주는 방식으로 코드를 짜면 더 깔끔해질 수 있겠다는 생각이 들었습니다!
```java
Piece white = add(Piece.WHITE_COLOR);
        verifySize(1);
        verifyFindPawn(white, 0);
```
유닛 테스트 환경에 한해서는 예상되는 값들을 직접적으로 넣어주어도 좋겠군요..! 
좋은 코드 잘 참고하였습니다 감사합니다 @Polynomeer

ksundong commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Board {
+
+    ArrayList<Pawn> pawns;
```
이 코드는 **접근 제어자**가 없네요~ 접근제어자에 대해서 좀 더 신경써주세요!

- 어떤 접근 제어자를 왜 붙여주어야 할까요?

ksundong replied:
보통 Java에서는 `ArrayList`같은 구체 클래스보다는 `List`와 같은 인터페이스 내지는 추상 클래스를 선언해줍니다.

- **다형성**에 대해서 한 번 공부해보세요. 사실 구체적인 클래스를 적어줘도 아무런 문제가 되지 않지만, 만약 `LinkedList`와 같은 다른 구체클래스로 변경이 필요한 경우 이 필드의 타입도 변경되어야 할 수 있습니다.

jihye-woo replied:
### 접근 제어자
OMG 왜 제가 private을 붙여주지 않았을까요...!!

### 선언 타입은 List로
마침 #46 에서도 [언급](https://github.com/codesquad-members-2021/java-chess/pull/46#discussion_r573748561)된 부분이라서 참고해서 공부하였습니다!

감사합니다!!😆😆

ksundong commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -0,0 +1,29 @@
+
+    ArrayList<Pawn> pawns;
+
+    public Board() {
+        this.pawns = new ArrayList<>();
+    }
```
이 코드는 굳이 기본 생성자를 정의하고, 그 안에서 할당하는 것 보다는 그냥 필드에 할당을 시켜줘도 괜찮을 것 같습니다.

jihye-woo replied:
Board 생성 시 Pawn을 담는 부분은 ArrayList로 선언해준다고 요구사항에 명시되어있기도 하고, pawns가 다른 구현체로 바뀔 일은 거의 발생하지 않을테니까, 필드에서 바로 초기화 시켜주어도 좋겠군요! 
감사합니다🙇‍♀️

ksundong commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public Pawn findPawn(int idx) {
+        if (idx < 0 || size() < idx) {
+            throw new IndexOutOfBoundsException("Index number " + idx + " is out of range! There is/are " + size() + " pieces on the board.");
+        }
```
예외처리를 잘 하셨습니다. 💯

ksundong replied:
한가지 의문점이 드는건 `idx`가 `size()`와 같을 때, 원하는 대로 동작하지 않을 것 같아요.

예를 들어  `pawns`에 아무 요소가 없는 경우. `findPawn(0)`을 수행하면, 해당 조건문을 만족하지 않아서, 통과하게 될 것 같습니다.
이런 엣지 케이스를 발견했다면, 테스트 코드를 추가해서 해당 문제가 재차 발생하지 않도록 방지해주세요.
이런 경우는 경계 조건에 해당하는데, 이런 경계 값은 항상 주의해서 작성해주셔야 합니다.

jihye-woo replied:
OMG... 그렇군요 이런 예외 케이스들을 잘 걸러내려고 TDD를 연습하는건데 정작 꼭 써야할 부분에다가 활용을 제대로 못했네요😂😂 주의하겠습니다! 코멘트 주신 부분 참고해서 예외 상황도 처리했습니다!

자세하게 봐주셔서 감사합니다!🙇‍♀️

---

그런데 한 가지 **궁금한 사항**이 있습니다
현재 테스트 메소드를 아래와 같이 작성하였습니다.
```java
@Test
    @DisplayName("보드에 없는 폰을 찾으려고 할 때, 에러가 발생해야 한다.")
    void find_보드에_없는_폰을_찾는다() {
        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            board.findPawn(0);
        });
    }
```
문제는 아래와 같이 `size() == idx`인 경우를 걸러내지 못하는 경우에도 똑같이 `IndexOutOfBoundsException`을 발생시켜서 테스트가 통과하는 상황이 생깁니다.
``` java
public Pawn findPawn(int idx) {
        if (idx < 0 || size() < idx) {
            throw new IndexOutOfBoundsException("Index number " + idx + " is out of range! There is/are " + size() + " pieces on the board.");
        }
        return pawns.get(idx);
    }
```

이런 경우에는 어떤 식으로 테스트 메소드를 작성해서 위와 같이  `size() == idx`인 경우를 놓치는 상황을 방지할 수 있을까요..?!
사실 제가 예외 처리를 해준 부분이 그저 예외 상황이 발생하게 되었을 때 구체적인 메세지를 추가하는 기능 정도밖에 되지 않아서, 이런 경우는 어떻게 유닛 테스트로 사전에 핸들링해줄 수 있을지 감이 오질 않습니다..!😂

너무 구체적이고 직접적으로 여쭤보는 것 같아서 죄송합니당😂

ksundong replied:
일단 AssertJ의 `assertThatThrownby()`로 바꿔주세요.
[이 글](https://www.baeldung.com/assertj-exception-assertion)을 참고해보세요.
힌트는 말씀하셨듯이 메시지가 서로 다릅니다!

죄송하실거 없어요! 궁금한건 여쭤보셔도 됩니다.

jihye-woo replied:
<img src="https://user-images.githubusercontent.com/74452069/107843259-fab8a580-6e0c-11eb-826b-e0a39eed56e4.png" width="80%">
오왕 같은 예외 클래스라도 에러 메세지로 따로 잡아낼 수 있군요!😮😮
에러 메세지가 바뀌는 경우에 따로 수정을 해줘야하는 문제가 있을 것 같아서 조심히 사용해주어야겠습니당😂

@ksundong 덕분에 하나 더 알아갑니다 감사합니다!!

ksundong replied:
대개의 경우는 에러 메시지를 바꿀일이 없다고 보는 것이 맞겠죠?
또한 바뀌더라도 테스트가 실패하니 테스트를 고쳐주면 됩니다.
테스트는 성공하면 안되는 테스트가 성공하는 것이 위험한 것이지, 성공해야하는 테스트가 실패하는 것은 고쳐주면 되니 괜찮습니다.
다만, 테스트 코드를 수정해야한다는 것은 조금 비용이 들기 때문에, 이를 어떻게 해결할지 고민해보는 것도 좋겠네요.

ksundong commented on src/main/java/net/honux/pieces/Pawn.java:
```diff
@@ -0,0 +1,23 @@
+package net.honux.pieces;
+
+public class Pawn {
+    public static final String WHITE_COLOR = "white";
+    public static final String BLACK_COLOR = "black";
+
```
👍

ksundong commented on src/main/java/net/honux/pieces/Pawn.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public Pawn(String color) {
+        if (!color.equals(WHITE_COLOR) && !color.equals(BLACK_COLOR)) {
+            throw new IllegalArgumentException("Pawn color should be 'white' or 'black'!");
+        }
```
예외처리 역시 잘 해주셨어요.

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,32 @@
+import net.honux.pieces.Pawn;
+import org.junit.jupiter.api.*;
+
+public class BoardTest {
```
JUnit5에서는 `public` 키워드를 테스트 클래스에도 굳이 안 붙여줘도 됩니다.

jihye-woo replied:
앗 테스트 '메소드'에만 신경을 썼네요ㅠ 감사합니다!

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Test
+    @DisplayName("보드 위에 폰 추가하고 추가된 폰이 보드에 정상적으로 반영되어야 한다.")
+    void add_폰이_보드에_정상적으로_추가되는지_확인() throws Exception {
```
저는 가급적이면 메서드명을 한글을 쓸 것이라면 `DisplayName`은 없애도 된다고 생각해요.
반면에 `DisplayName`을 쓸 것이라면 메서드 명은 영문으로 적어주는 편입니다.

jihye-woo replied:
그렇군요! DispalyName이 이미 다 설명해주고 있으니 상관없겠군요
Problems에 `Non-ASCII characters'라고 뜨는 부분이 신경쓰이기도 했는데 수정해야습니당😂 
감사합니다!

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,32 @@
+        addPawnAndCheckTheBoard(white);
+
+        Pawn black = new Pawn(Pawn.BLACK_COLOR);
+        addPawnAndCheckTheBoard(black);
```
꼭 굳이 두 색상의 Pawn을 추가해서 확인해줄 필요가 있을까요~?
테스트를 할 때, 목적을 명확히 해야합니다.

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,32 @@
+        int sizeBeforeAddThePawn = board.size();
+        board.add(pawn);
+        Assertions.assertEquals(sizeBeforeAddThePawn + 1, board.size());
+        Assertions.assertEquals(pawn, board.findPawn(sizeBeforeAddThePawn));
```
29번 라인은 읽기 번거로워진다는 문제가 있습니다. (왜 이전 사이즈를 가져와서 찾지?라는 의문이 들 수 있어요.)
중복을 제거하는 것도 좋지만, 저라면 그냥 변수 하나를 더 할당해서 더 읽기 좋은 코드를 만들어 주었을 것 같네요.

ksundong commented on src/test/java/net/honux/pieces/MainTest.java:
```diff
@@ -0,0 +1,36 @@
+        String unsupported_color = "red";
+        Assertions.assertThrows(IllegalArgumentException.class, () -> {
+            Pawn redPawn = new Pawn(unsupported_color);
+        });
```
그냥 이런건 변수로 안빼주셔도 돼요 ㅎㅎ
오히려 변수명이 snake case가 되어버리는 불상사가 발생해버렸습니다.

ksundong commented on 미션2.md:
```diff
@@ -0,0 +1,123 @@
+### 중복 제거
+
+- 지금까지 구현한 코드에서 중복되는 부분이 있는지 확인하고 제거한다.
+- 프로덕션 코드 뿐만 아니라 테스트 코드에서도 중복이 발생하는지 확인하고, @Before 애노테이션을 활용해 중복을 제거한다.
```
이번에도 잘 작성해 주셨네요! 👍 💯 🥇

ksundong commented on src/test/java/net/honux/pieces/MainTest.java:
```diff
@@ -0,0 +1,36 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+class PawnTest {
```
파일명이랑 클래스명이랑 다르네요. 수정해주세요!

jihye-woo replied:
#29 미션 1때도 PR 내용 중 `요구사항과 맞지 않습니다!` 에서 말씀 드렸던 부분입니다..!ㅠㅠ

PawnTest.java라는 이름으로 java 파일을 생성하게 되면 java 파일로 인식을 못하는 오류가 계속 발생해서
현재 MainTest.java에 내부클래스 형태로 PawnTest 클래스를 구현한 상태입니다.

다시 프로젝트를 clone해와서 새로 시작해보기도 했지만, 결국 해결하지 못했습니다..ㅠ

ksundong replied:
흠... 저는 잘 되는데 무슨일일까요?

jihye-woo replied:
디온 천사...👼 덕분에 해결했어요 감사합니다😆

ksundong reviewed (changes requested):
안녕하세요! Woody 벌써 두 번째 리뷰네요!! 잘 하고 계십니다.
이번에는 제가 예외 상황을 발견해서, 변경요청을 드리고자 합니다. 그 밖에 리뷰와 코멘트를 남겼습니다.
커밋메시지와 PR은 정말 잘 작성해주셨어요. 💯 

## 리뷰

- 요구사항은 과하게(일단 해보시고, 판단해 드리겠습니다.)는 말고 어느정도 더 발전된 방향으로 하시는 것을 권장합니다. `enum` 사용하셔도 돼요. 요구사항에 맞추기 위해 코드를 작성하지 않으셔도 됩니다.
- 컴파일 에러가 발생하는 코드가 있나요? 아마 작업하시면서 사라진 것 같습니다.

---

- **접근 제어자**에 유의해주세요!!(필수)
- **다형성**에 대해서 한 번 학습해보세요.
- 예외 케이스를 수정해주세요.(필수)
- JUnit5에는 `public` 접근 제어자를 '테스트 클래스' 또는 '테스트 메서드'에 굳이 붙여주지 않으셔도 됩니다.
- 가급적이면 DisplayName을 사용하신다면 메서드명은 영어로 적어주시는게 낫다고 생각해요. 아니면 메서드 명만 한글로 작성해주시면 된다고 생각해요. 왜냐하면 제가 보기엔 같은 말을 두 번 반복하는 느낌이거든요.
- 테스트 코드는 목적이 중요합니다. 어떤 것을 테스트할 것인지 판단하고, 그에 대해서만 판단해주세요.
- 파일명과 클래스명은 꼭 일치시켜주세요.

고생 많으셨습니다. Woody도 새해 복 많이 받으세요!
변경 요청 드린 부분 수정해주시면 다시 리뷰 하겠습니다~

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,13 @@
+    @Test
+    public void create() throws Exception {
+
+    }
```
요런 빈 Test는 항상 성공하기 때문에 `fail()` 같은 메서드를 넣어서 실패하게끔 만들어주셔도 좋습니다.
그리고 지금 봤는데 `throws Exception`이 왜 붙어있는지 궁금하네요~

jihye-woo replied:
### Fail()
그렇군요! @ksundong 말씀 참고해서 아예 Live Template에 test 키워드로 테스트 메소드를 자동 생성하면 Body 부분에 `org.junit.jupiter.api.Assertions.fail(\"Not implemented\");` 가 자동으로 생성되게 만들어놓았습니다!

```java
 @Test
    @DisplayName("Test name")
    void method_name() {
        Assertions.fail("Not implemented");
    }
```

### throws Exception
이 부분은 제가 요구사항 코드를 참고하는 과정에서 넣게 되었는데 다음부터는 요구사항 코드를 100% 따라하는게 아니라 흐름만 보고 제 코드로 만드는 연습을 해야할 것 같네요..😂

좋은 코멘트 감사합니다!

ksundong replied:
와 Live Template까지 만드시다니 대단하시네요. 👍👍

jihye-woo committed:
docs : 코드리뷰 내용 정리

jihye-woo committed:
refactor : pawns 초기화문 수정 (Board 클래스 내부)

  - 접근 지시자 private 수정
  - List<Pawn>으로 업캐스팅

jihye-woo committed:
refactor : BoardTest 수정

  > 메소드명 영어로 변경

  > 작은 기능 단위로 메소드 분할
    - addPawn()
    - checkTheBoard()

  > 테스트 클래스 접근 제어자
    - public → default 변경

jihye-woo committed:
refactor : MainTest 수정

  > 메소드명 영어로 변경

  > assertAll() 사용
    - white pawn 생성 여부와 관계없이 black pawn이 생성되는지 확인

  > 불필요한 변수 제거

  > assertThat → assertEquals로 변경
    - 이미 사용하고 있는 Assertion 구문으로 통일

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,32 @@
+    private void addPawnAndCheckTheBoard(Pawn pawn) {
+        int sizeBeforeAddThePawn = board.size();
+        board.add(pawn);
+        Assertions.assertEquals(sizeBeforeAddThePawn + 1, board.size());
+        Assertions.assertEquals(pawn, board.findPawn(sizeBeforeAddThePawn));
```
여기두 AssertJ로 변경해주세요~

jihye-woo replied:
assertEquals 대신 AssertJ의 assertThat으로 변경하였습니다!
감사합니다😆

jihye-woo committed:
refactor : MainTest 클래스명 변경

  > 클래스명 변경
    - MainTest.java → PawnTest.java

jihye-woo committed:
fix : 경계 조건 예외 수정

  > size() == idx
    - idx가 사이즈와 같을 때도 고려

  > 예외 상황 세분화 - 음수 idx 따로 처리
    - NegativeArraySizeException 추가
    - 테스트 케이스 추가

jihye-woo committed:
refactor : assertJ 사용

  > assertEqual 대신 assertThat 사용

ksundong commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Board {
+
+    private final List<Pawn> pawns = new ArrayList<>();
```
오 `final`을 써주셨네요! 혹시 `final`을 써주신 이유가 있을까요?

jihye-woo replied:
변경할 필요가 없는 필드 값은 `final`로 설정해서 변경 가능성을 없애는 것이 좋을 것 같아 사용해보았습니다..!

ksundong replied:
필드 자체의 변경은 안되지만, 이런 컬렉션이나 객체는 내부의 상태가 변경될 수 있음도 생각해주세요!!

ksundong commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public Pawn findPawn(int idx) {
+        if (idx < 0) {
+            throw new NegativeArraySizeException("Negative numeric value " + idx + " cannot be a idx!");
+        }
+        if (size() <= idx) {
+            throw new IndexOutOfBoundsException("Index number " + idx + " is out of range!");
+        }
```
둘을 구체적인 예외를 발생시켜주기 위해서 예외를 나눈 것 같은데
`NegativeArraySizeExeption`이 맞을까요? 이 예외는 뭔가 초기화 시에 발생할 것 같아요.
차라리 저는 둘 다 `IndexOutOfBoundsExeption`이 되는 편이 나을 것 같습니다!

jihye-woo replied:
넵 생각해보니 굳이 `NegativeArraySizeExeption`을 사용할 필요가 없었던 것 같습니다! 
다시 수정해야겠습니다 감사합니다!😊

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,57 @@
+    @Test
+    @DisplayName("보드에 있는 폰을 정상적으로 찾을 수 있어야 한다.")
+    void findPawnOnTheBoard() {
+        Pawn firstPawn = new Pawn();
+        addThePawnProperly(firstPawn, 0);
+
+        Pawn secondPawn = new Pawn();
+        addThePawnProperly(secondPawn, 1);
```
일단 테스트 환경 준비 → 수행 → 테스트 단정(assertion)의 순서로 테스트를 작성해주시는게 테스트가 어떤 동작을 테스트하려는지 명확해집니다.

지금 하고자 하는 것은, 보드에 있는 폰을 정상적으로 찾을 수 있는 것인지 알아내는 것인데, 불필요하게 테스트하는 항목들이 너무 많습니다.

나머지 기능은 정상적으로 동작한다 가정하고, 되도록이면 테스트 하려는 대상만 테스트해주세요!

jihye-woo replied:
요구사항에 있는 부분을 최대한 반영해서 일단은 Pawn을 두 번 추가해서 두 번 검증해주었습니다..!

ksundong commented on src/test/java/net/honux/chess/BoardTest.java:
```diff
@@ -0,0 +1,57 @@
+    @Test
+    @DisplayName("보드에 없는 폰을 찾으려고 할 때, 에러가 발생해야 한다.")
+    void findPawnNotOnTheBoard() {
+        assertThatThrownBy(() -> board.findPawn(0))
+                .isInstanceOf(IndexOutOfBoundsException.class)
+                .hasMessageContaining("Index number 0 is out of range!");
```
👍 
요건 보드에 폰이 존재하지 않을 때, 예외가 발생한다는 것이 맞을 것 같아요.
그리고 자바의 기본예외를 사용하는 것도 좋지만, 표현하기 어려우면 새로운 예외 정의를 하시는 방안도 생각해보세요!

jihye-woo replied:
"보드에 폰이 존재하지 않을 때, 폰을 찾으려고 시도하면 에러가 발생해야 한다." 로 수정하였습니다!
지금은 새로운 예외를 정의할 수 있다는 것만 알아두고 나중에 기회가 되면 적용해보도록 하겠습니다! 감사합니다🙇‍♀️

ksundong commented on src/test/java/net/honux/pieces/PawnTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    @DisplayName("흰색 폰과 검은색 폰이 생성되어야 한다")
+    void createWhiteAndBlackPawns() {
+        assertAll(
+                () -> verifyPawn(Pawn.WHITE_COLOR),
+                () -> verifyPawn(Pawn.BLACK_COLOR));
+    }
```
👍

ksundong commented on src/test/java/net/honux/pieces/PawnTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    @DisplayName("다른 색 폰이 생성되면 안된다")
+    void createRedPawn() {
+        assertThrows(IllegalArgumentException.class, () -> {
+            Pawn redPawn = new Pawn("red");
+        });
```
요거 `assertThatThrownBy()`로 고쳐주시면 좋을 것 같습니다!
assertJ를 사용하는 편이 전 좋다고 생각해요!

jihye-woo replied:
assertJ로 변경하는 과정에서 저 부분은 생각을 못했었네요 수정하겠습니다! 
감사합니다😆

ksundong commented on 미션2.md:
```diff
@@ -0,0 +1,154 @@
+    - 테스트 메소드 책임 분리 (동사 단위로 분리)
+    - 목적에 맞는 테스트 로직 작성 (굳이 두 색상을 추가해서 확인?)
+
+7. PawnTest 클래스
+    - MainTest -> PawnTest 이름 변경
+    - 해야하지만... 어떻게 해야할지 모르겠다.
```
되신거 아닌가용?

jihye-woo replied:
앗 수정을 안했군요..! 수정하겠습니당

ksundong reviewed (changes requested):
명절에도 고생하셨습니다!
테스트 코드를 좀 더 테스트 하고자 하는 것이 명확해지게 작성해보시는 연습을 요청드립니다!

jihye-woo committed:
refactor : assertJ 사용

  > assertThrows 대신 assertThatThrownBy 사용

jihye-woo committed:
refactor : BoardTest 메소드 수정

  > findPawnNotOnTheBoard()
    - DisplayName 수정

  > addThePawnProperly()
    - addPawn()과 checkTheBoard()를 하나로 묶어줌

jihye-woo committed:
fix : 경계 조건에 대한 예외 클래스 변경

  > Board findPawn()
    - ArrayIndexOutOfBoundsException로 변경

  > BoardTest
    - findPawn() 테스트 메소드 모두 ArrayIndexOutOfBoundsException를 핸들하도록 변경

ksundong reviewed (approved):
고생하셨습니다! 다음단계 진행해주셔도 좋을 것 같아요! 👍
